tests/backends/console/arduino/test_uno.py

```diff
@@ -0,0 +1,228 @@
+
+    with pytest.raises(ValueError):
+        backend.get_gpio_pin_digital_state(4)
+
+
+def test_read_gpio_pin_digital_state() -> None:
+    """Test that we can read the digital state of a pin."""
+    backend = ArduinoUnoConsoleBackend(
+        "TestBoard",
+        console_class=MockConsole,
+    )
+
+    backend._pins[10].mode = GPIOPinMode.DIGITAL_INPUT
+    backend._console.next_input = "True"  # type: ignore
+    assert backend.read_gpio_pin_digital_state(10)
+
+
+def test_read_gpio_pin_digital_state_bad_mode() -> None:
+    """Test that we cannot read a digital state in the wrong mode."""
+    backend = ArduinoUnoConsoleBackend(
+        "TestBoard",
+        console_class=MockConsole,
+    )
+
+    backend._console.next_input = "True"  # type: ignore
+
+    with pytest.raises(ValueError):
+        assert backend.read_gpio_pin_digital_state(10)
+
+
+def test_read_gpio_pin_analogue_value() -> None:
+    """Test that we can read an analogue value."""
+    backend = ArduinoUnoConsoleBackend(
+        "TestBoard",
+        console_class=MockConsole,
+    )
+
+    backend._console.next_input = "4.3"  # type: ignore
+    backend._pins[5].mode = GPIOPinMode.ANALOGUE_INPUT
+    assert backend.read_gpio_pin_analogue_value(5) == 4.3
+
+
+def test_read_gpio_pin_analogue_value_bad_mode() -> None:
+    """Test that we cannot read an analogue value in the wrong mode."""
+    backend = ArduinoUnoConsoleBackend(
+        "TestBoard",
+        console_class=MockConsole,
+    )
+
+    backend._console.next_input = "4.3"  # type: ignore
+    with pytest.raises(ValueError):
+        assert backend.read_gpio_pin_analogue_value(5) == 4.3
+
+
+def test_write_gpio_pin_dac_value() -> None:
+    """Test that this isn't implemented."""
+    backend = ArduinoUnoConsoleBackend("test")
+
+    with pytest.raises(NotImplementedError):
+        backend.write_gpio_pin_dac_value(10, 1.0)
+
+
+def test_write_gpio_pin_pwm_value() -> None:
+    """Test that this isn't implemented."""
+    backend = ArduinoUnoConsoleBackend("test")
+
+    with pytest.raises(NotImplementedError):
+        backend.write_gpio_pin_pwm_value(10, 1.0)
+
+
+def test_get_led_state() -> None:
+    """Test that we can get the LED state."""
+    backend = ArduinoUnoConsoleBackend(
+        "TestBoard",
+        console_class=MockConsole,
+    )
+    assert not backend.get_led_state(0)
+
+    backend._pins[13].digital_state = True
+    assert backend.get_led_state(0)
+
+
+def test_set_led_state() -> None:
+    """Test that we can set the LED state."""
+    backend = ArduinoUnoConsoleBackend(
+        "TestBoard",
+        console_class=MockConsole,
+    )
+
+    backend._console.expects = "Set pin 13 state to False"  # type: ignore
+    backend.set_led_state(0, False)
+    assert not backend.get_led_state(0)
+
+    backend._console.expects = "Set pin 13 state to True"  # type: ignore
+    backend.set_led_state(0, True)
+    assert backend.get_led_state(0)
```

Comment:
Could we make this check the LED and digital pin 13 are the same thing? i.e. write using `set_led_state(0, ...)` and read back using `get_gpio_pin_digital_state(13)`, and vice versa.

Reply:
b7e8229